File: DXDAL/ProjectDetails.vb
Imports System
Imports DevExpress.Xpo
Imports DevExpress.Data.Filtering

Namespace SPGData

    Public Class ProjectDetails
        Inherits AuditableBase

        ' Fields...
        Private _allowOverrun As Boolean
        'Private _percentageComplete As Decimal
        'Private _palletsProduced As Integer
        'Private _bagsProduced As Integer
        'Private _unitsProduced As Integer
        Private _palletsRequested As Integer
        Private _bagsRequested As Integer
        Private _unitsRequested As Integer
        Private _projectItem As Items
        Private _customerPO As String
        Private _project As Project

        <Association("Project-Details")>
        Public Property Project As Project
            Get
                Return _project
            End Get
            Set(ByVal Value As Project)
                SetPropertyValue(Of Project)("Project", _project, Value)
            End Set
        End Property

        Public Property CustomerPO As String
            Get
                Return _customerPO
            End Get
            Set(ByVal Value As String)
                SetPropertyValue(Of String)("CustomerPO", _customerPO, Value)
            End Set
        End Property

        Public Property ProjectItem As Items
            Get
                Return _projectItem
            End Get
            Set(ByVal Value As Items)
                SetPropertyValue(Of Items)("ProjectItem", _projectItem, Value)
            End Set
        End Property

        Public Property UnitsRequested As Integer
            Get
                Return _unitsRequested
            End Get
            Set(ByVal Value As Integer)
                SetPropertyValue(Of Integer)("UnitsRequested", _unitsRequested, Value)
            End Set
        End Property

        Public Property BagsRequested As Integer
            Get
                Return _bagsRequested
            End Get
            Set(ByVal Value As Integer)
                SetPropertyValue(Of Integer)("BagsRequested", _bagsRequested, Value)
            End Set
        End Property

        Public Property PalletsRequested As Integer
            Get
                Return _palletsRequested
            End Get
            Set(ByVal Value As Integer)
                SetPropertyValue(Of Integer)("PalletsRequested", _palletsRequested, Value)
            End Set
        End Property

        Public Property AllowOverrun As Boolean
            Get
                Return _allowOverrun
            End Get
            Set(ByVal Value As Boolean)
                SetPropertyValue(Of Boolean)("AllowOverrun", _allowOverrun, Value)
            End Set
        End Property
        
        <Association("Production-ProjectDetail", GetType(ProductionProjectDetails)), Aggregated()>
        Public ReadOnly Property ProjectProduction As XPCollection(Of ProductionProjectDetails)
            Get
                Return GetCollection(Of ProductionProjectDetails)("ProjectProduction")
            End Get
        End Property

        <Association("Project-ProductionSchedule", GetType(ProductionSchedule)), Aggregated()>
        Public ReadOnly Property ProjectProductionSchedule As XPCollection(Of ProductionSchedule)
            Get
                Return GetCollection(Of ProductionSchedule)("ProjectProductionSchedule")
            End Get
        End Property

        <PersistentAlias("IsNull(ProjectProduction.Sum(UnitsProduced), 0)")>
        Public ReadOnly Property UnitsProduced As Integer
            Get
                Return Convert.ToInt32(EvaluateAlias("UnitsProduced"))
            End Get
        End Property

        <PersistentAlias("IsNull(ProjectProductionSchedule.Sum(Quantity), 0)")>
        Public ReadOnly Property UnitsScheduled As Integer
            Get
                Return Convert.ToInt32(EvaluateAlias("UnitsScheduled"))
            End Get
        End Property

        'used to identify if there's still units needed to produce
        Public ReadOnly Property UnitsNeeded As Integer
            Get
                Return Math.Max(UnitsRequested - UnitsProduced, 0)
            End Get
        End Property

        'Used to identify if there's still units needed to be scheduled
        Public ReadOnly Property UnitsNeededToSchedule As Integer
            Get
                Return Math.Max(UnitsRequested - Math.Max(UnitsProduced, UnitsScheduled), 0)
            End Get
        End Property

        <PersistentAlias("ProjectProduction.Sum(BagsProduced)")>
        Public ReadOnly Property BagsProduced As Integer
            Get
                Return Convert.ToInt32(EvaluateAlias("BagsProduced"))
            End Get
        End Property

        <PersistentAlias("ProjectProduction.Sum(PalletsProduced)")>
        Public ReadOnly Property PalletsProduced As Integer
            Get
                Return Convert.ToInt32(EvaluateAlias("PalletsProduced"))
            End Get
        End Property

        Public ReadOnly Property PercentageComplete As Double
            Get
                If UnitsProduced = 0 OrElse UnitsRequested = 0 Then
                    Return 0
                End If

                Return UnitsProduced / UnitsRequested * 100
            End Get
        End Property

        Public ReadOnly Property EstimatedShifts As Double?
            Get
                If ProjectItem Is Nothing OrElse UnitsRequested = 0 OrElse ProjectItem.ItemProdStandard.HasValue = False OrElse ProjectItem.ItemProdStandard.Value = 0 Then
                    Return Nothing
                End If

                Return (UnitsRequested * ProjectItem.intQtyPerUnit) / (ProjectItem.ItemProdStandard.Value * 358.3333)     '7:10 * 50 min per hour
            End Get
        End Property

        Public Sub New()
            MyBase.New()
            ' This constructor is used when an object is loaded from a persistent storage.
            ' Do not place any code here.			
        End Sub

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
            ' This constructor is used when an object is loaded from a persistent storage.
            ' Do not place any code here.			
        End Sub

        Public Overrides Sub AfterConstruction()
            MyBase.AfterConstruction()
            ' Place here your initialization code.
        End Sub

        Public Shadows Class FieldsClass
            Inherits DevExpress.Xpo.PersistentBase.FieldsClass
            Public Sub New()
                MyBase.New()
            End Sub
            Public Sub New(ByVal propertyName As String)
                MyBase.New(propertyName)
            End Sub
            Public ReadOnly Property Oid() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("Oid"))
                End Get
            End Property
            Public ReadOnly Property Project() As Project.FieldsClass
                Get
                    Return New Project.FieldsClass(GetNestedName("Project"))
                End Get
            End Property
            Public ReadOnly Property CustomerPO() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("CustomerPO"))
                End Get
            End Property
            Public ReadOnly Property ProjectItem() As Items.FieldsClass
                Get
                    Return New Items.FieldsClass(GetNestedName("ProjectItem"))
                End Get
            End Property
            Public ReadOnly Property UnitsRequested() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("UnitsRequested"))
                End Get
            End Property
            Public ReadOnly Property BagsRequested() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("BagsRequested"))
                End Get
            End Property
            Public ReadOnly Property PalletsRequested() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("PalletsRequested"))
                End Get
            End Property
            Public ReadOnly Property UnitsProduced() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("UnitsProduced"))
                End Get
            End Property
            Public ReadOnly Property BagsProduced() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("BagsProduced"))
                End Get
            End Property
            Public ReadOnly Property PalletsProduced() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("PalletsProduced"))
                End Get
            End Property
            Public ReadOnly Property PercentageComplete() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("PercentageComplete"))
                End Get
            End Property
            Public ReadOnly Property AllowOverrun() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("AllowOverrun"))
                End Get
            End Property
            Public ReadOnly Property EstimatedShifts() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("EstimatedShifts"))
                End Get
            End Property
            Public ReadOnly Property UnitsScheduled() As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("UnitsScheduled"))
                End Get
            End Property
            Public ReadOnly Property UnitsNeeded As OperandProperty
                Get
                    Return New OperandProperty(GetNestedName("UnitsNeeded"))
                End Get
            End Property
        End Class
        Private Shared _fields As FieldsClass
        Public Shared Shadows ReadOnly Property Fields() As FieldsClass
            Get
                If ReferenceEquals(_fields, Nothing) Then
                    _fields = New FieldsClass()
                End If
                Return _fields
            End Get
        End Property

    End Class
End Namespace

File: DXDAL/ProjectDetails.cs
using System;
using DevExpress.Xpo;
using DevExpress.Data.Filtering;

namespace DXDAL
{
    namespace SPGData
    {
        public class ProjectDetails : AuditableBase
        {

            // Fields...
            private bool _allowOverrun;
            // Private _percentageComplete As Decimal
            // Private _palletsProduced As Integer
            // Private _bagsProduced As Integer
            // Private _unitsProduced As Integer
            private int _palletsRequested;
            private int _bagsRequested;
            private int _unitsRequested;
            private Items _projectItem;
            private string _customerPO;
            private Project _project;

            [Association("Project-Details")]
            public Project Project
            {
                get => _project;
                set => SetPropertyValue("Project", ref _project, value);
            }

            public string CustomerPO
            {
                get => _customerPO;
                set => SetPropertyValue("CustomerPO", ref _customerPO, value);
            }

            public Items ProjectItem
            {
                get => _projectItem;
                set => SetPropertyValue("ProjectItem", ref _projectItem, value);
            }

            public int UnitsRequested
            {
                get => _unitsRequested;
                set => SetPropertyValue<int>("UnitsRequested", ref _unitsRequested, value);
            }

            public int BagsRequested
            {
                get => _bagsRequested;
                set => SetPropertyValue<int>("BagsRequested", ref _bagsRequested, value);
            }

            public int PalletsRequested
            {
                get => _palletsRequested;
                set => SetPropertyValue<int>("PalletsRequested", ref _palletsRequested, value);
            }

            public bool AllowOverrun
            {
                get => _allowOverrun;
                set => SetPropertyValue("AllowOverrun", ref _allowOverrun, value);
            }

            [Association("MO-ProjectDetail", typeof(MO))]
            [Aggregated()]
            public XPCollection<MO> ProjectMO => GetCollection<MO>("ProjectMO");

            [Association("Production-ProjectDetail", typeof(ProductionProjectDetails))]
            [Aggregated()]
            public XPCollection<ProductionProjectDetails> ProjectProduction => GetCollection<ProductionProjectDetails>("ProjectProduction");

            [Association("Project-ProductionSchedule", typeof(ProductionSchedule))]
            [Aggregated()]
            public XPCollection<ProductionSchedule> ProjectProductionSchedule => GetCollection<ProductionSchedule>("ProjectProductionSchedule");

            [PersistentAlias("IsNull(ProjectProduction.Sum(UnitsProduced), 0)")]
            public int UnitsProduced => Convert.ToInt32(EvaluateAlias("UnitsProduced"));

            [PersistentAlias("IsNull(ProjectProductionSchedule.Sum(Quantity), 0)")]
            public int UnitsScheduled => Convert.ToInt32(EvaluateAlias("UnitsScheduled"));

            // used to identify if there's still units needed to produce
            public int UnitsNeeded => Math.Max(UnitsRequested - UnitsProduced, 0);

            // Used to identify if there's still units needed to be scheduled
            public int UnitsNeededToSchedule => Math.Max(UnitsRequested - Math.Max(UnitsProduced, UnitsScheduled), 0);

            [PersistentAlias("ProjectProduction.Sum(BagsProduced)")]
            public int BagsProduced => Convert.ToInt32(EvaluateAlias("BagsProduced"));

            [PersistentAlias("ProjectProduction.Sum(PalletsProduced)")]
            public int PalletsProduced => Convert.ToInt32(EvaluateAlias("PalletsProduced"));

            public double PercentageComplete
            {
                get
                {
                    if (UnitsProduced == 0 || UnitsRequested == 0)
                        return 0;

                    return UnitsProduced / (double)UnitsRequested * 100;
                }
            }

            public double RemainingBalance => UnitsRequested - UnitsProduced;

            public double? EstimatedShifts
            {
                get
                {
                    if (ProjectItem == null || RemainingBalance == 0 || ProjectItem.ItemProdStandard.HasValue == false || ProjectItem.ItemProdStandard.Value == 0)
                        return default(double?);

                    return (RemainingBalance * ProjectItem.intQtyPerUnit) / (ProjectItem.ItemProdStandard.Value * 358.3333);     // 7:10 * 50 min per hour
                }
            }

            public ProjectDetails() : base()
            {
            }

            public ProjectDetails(Session session) : base(session)
            {
            }

            public override void AfterConstruction()
            {
                base.AfterConstruction();
            }

            public new class FieldsClass : PersistentBase.FieldsClass
            {
                public FieldsClass() : base()
                {
                }
                public FieldsClass(string propertyName) : base(propertyName)
                {
                }
                public OperandProperty Oid => new OperandProperty(GetNestedName("Oid"));
                public Project.FieldsClass Project => new Project.FieldsClass(GetNestedName("Project"));
                public OperandProperty CustomerPO => new OperandProperty(GetNestedName("CustomerPO"));
                public Items.FieldsClass ProjectItem => new Items.FieldsClass(GetNestedName("ProjectItem"));
                public OperandProperty UnitsRequested => new OperandProperty(GetNestedName("UnitsRequested"));
                public OperandProperty BagsRequested => new OperandProperty(GetNestedName("BagsRequested"));
                public OperandProperty PalletsRequested => new OperandProperty(GetNestedName("PalletsRequested"));
                public OperandProperty UnitsProduced => new OperandProperty(GetNestedName("UnitsProduced"));
                public OperandProperty BagsProduced => new OperandProperty(GetNestedName("BagsProduced"));
                public OperandProperty PalletsProduced => new OperandProperty(GetNestedName("PalletsProduced"));
                public OperandProperty PercentageComplete => new OperandProperty(GetNestedName("PercentageComplete"));
                public OperandProperty AllowOverrun => new OperandProperty(GetNestedName("AllowOverrun"));
                public OperandProperty EstimatedShifts => new OperandProperty(GetNestedName("EstimatedShifts"));
                public OperandProperty UnitsScheduled => new OperandProperty(GetNestedName("UnitsScheduled"));
                public OperandProperty UnitsNeeded => new OperandProperty(GetNestedName("UnitsNeeded"));
            }
            private static FieldsClass _fields;
            public static new FieldsClass Fields
            {
                get
                {
                    if (ReferenceEquals(_fields, null))
                        _fields = new FieldsClass();
                    return _fields;
                }
            }
        }
    }
}
